### ISSUE TITLE
Where to get MBA5,2 ; MBP8,1 and MBP10,1 and what could be wrong

### ISSUE THREAD

**Issue by cedricbambooza** [gch1p/mmga#6]
Hi, I was trying to follow the instructions, but I can't find the Macbook Pro 10,1 or Macbook Air 5,2 on the Coreboot git.

I tried this git as suggested: 

`$ git clone --recurse-submodules https://review.coreboot.org/coreboot.git && cd coreboot`

What is the issue here? 

Thanks in advance.

**Comment by gch1p** [comment]
Hi, they are not merged to master. 

https://review.coreboot.org/c/coreboot/+/32604
https://review.coreboot.org/c/coreboot/+/32604

You'd need to pull and checkout to it. 

But mmga actually does it for you.

**Comment by cedricbambooza** [comment]
Thanks for the fast reply :) 

I understand and have to admit I didn't try mmga until now. 

Is there any good reason, why it is not merged to master?

**Comment by cedricbambooza** [comment]
Another question would be, when will those be ready for master?

**Comment by gch1p** [comment]
It's the long story. But since you asked... here it is.

(Disclaimer: everything below is just my opinion. I may be wrong. Or maybe not. Or maybe coreboot, as a project, at its current state, is just not for me.)

---

When I first contributed to coreboot, which was in 2018, it was easier: you upload a patch, maintainers would come, make some comments, you address them, fix stuff, and eventually it gets merged to master. It would take days, maybe weeks. As an example, my **very first** patch (I was really a newbie and barely even understood how it all ticks under the hood in the BIOS), the [ThinkPad W530 support](https://review.coreboot.org/c/coreboot/+/26136). Uploaded May 7, merged at Jun 6. A month. That's fine. I remember an old project member saying, "as long as it boots, it's good for merging". Ha-ha, I'd love that would still be true.

Then... I don't know what exactly happened or why. But starting maybe since 2019-2020 coreboot code review became too strict and even unbearable to my taste. You port a board, test it, everything works (or if something doesn't, you documenting it). Maintainers come, make some comments. You fix stuff. Then it turns out you must write a documentation page about how to flash the board, or else the patch will not be accepted. Okay, I took photos, even redacted them in Photoshop, wrote, I believe, good documentation, for all three macbooks I ported, and even [internal flashing guide](https://review.coreboot.org/c/coreboot/+/38770) as a bonus. Then somebody may come and tell you that some `#include`s are *not in alphabetical order* and you need to fix it. Or Kconfig options, *they should be in alphabetical order too*. Then somebody could point you to the fact that some spaces in some comments are not perfectly aligned. Or an **indentation in the commit message is not how maintainer likes it**, I'm not kidding. You fix it too, obviously it takes some time, not only fixing, but the review stages itself, because nobody can review immediately after you upload new patchset, people are busy with their stuff, which is actually normal, so days or even weeks may pass between these reviews, and with each new one you may receive some more "change this", "refactor that" pain. In the meantime, master changes, and then a person comes and says something like "let's rename these files and rewrite that code, it's not compatible with master now". You rebase, do it, and then... Well, you probably guessed.

It's not only with new boards. [This thing](https://review.coreboot.org/c/coreboot/+/37115) took more than a year to get accepted. It was rewritten at least two or three times, despite it was quite good from the beginning. Not perfect, okay, but it definitely worked and it could be fixed and merged faster. Actually, it barely has changed, I was mostly just moving functions here and there between files and changing minor stuff. Luckily I had patience to finish it as I find this patch quite important.

I could give more examples, but I think you get the idea.

There was once this "drama", as I called it then: https://www.reddit.com/r/coreboot/comments/lm7jh8/my_pain_with_documentation_contribution_in/. Back then, I did not completely agree with the guy. Hell, I even think I was that very reviewer who asked him to fix some stuff (100% wrong stuff, I must say, but still, it was the last straw for him) when he gave up and went on reddit. Today, **I agree with him completely**. Contributing to coreboot is a pain in the ass, and, for the moment, I'm done with it.

My work is mostly finished: the ports do work (if you have compatible RAM), it's all documented. In my blog you can find flashing guides with photos. I actually do use MBA 5,2 and MBP 8,1 with coreboot and linux and it works. MBP 10,1 works too but I don't use it, it's just lying around somewhere.

I may add support for new memory if people send dumps to me. Maybe, when I'll have the right mood, I'll rebase the patches again and fix some more stuff they're asking. Maybe I will even be lucky and the patch will receive a +2. (It all depends on who will review it.)

But, as of today, I can't tell when those will be ready for master. I haven't touch coreboot for months, but I'm still so sick of wasting my time for it. I think I spent 20% of time actually porting the boards, researching and doing all that interesting stuff, and 80% doing **nothing useful**.

---

Sorry if it was a bit emotional. If you will have any troubles with the ports, I'm here to help. I'm always ready to help anybody who is **actually going to use my code**, it's just the time-wasting coreboot review nonsense that I'm so sick of.

Now, about mmga. I didn't test it for quite some time. It pulls not only my coreboot patches, but also some stuff I don't control, such as SeaBIOS, and that stuff may break sometimes. Actually, SeaBIOS master was broken on MacBooks last time I tried it, but it was half a year ago or so.

If you have some experience with coreboot, BIOS flashing and all that stuff, I'd recommend building and flashing it manually and **externally**, without using mmga, because this way you have more understanding and control over the process. mmga may work, but it's fragile. Oh, and don't forget to check whether your RAM is supported.

**Comment by cedricbambooza** [comment]
I get the point :) 

From a coders perspective, starting to sort the code alphabetically sounds pretty stupid and like time wasted. 
Maybe also like chicane. 

I do appreciate your long and emotional answer. 

The last Maintainer I had to deal with, wrote something like "Tstststststststs", after I started to open fire to him because he left the repo unmaintained for weeks while something broke and he just had to merge a pull request. 
The issue got fixed in 5 days, the patch took nearly months. 
I was also preloaded for this because over 200 issues were open and some where 2 years old. 
Stuff like that triggers me. 

I think some maintainers don't go well with the job, maybe that's also the case with coreboot. 

I still hope it will be merged to master one day, but I think I will be ok with your patches till that day. 

You sound like you have sort of the right spirit behind the code.

**Comment by cedricbambooza** [comment]
I do have some questions left though: 

Is the sound on the MBA 5,2 headphone fixed/fixable? 

How about issues on the Macbook Pro 10,1? 
Are there any major? 

Sound on the headphones and thunderbolt are maginal for me. 

I will buy a MBA 4,2 soon, so I think that will work smoothly since the picky maintainers merged it a while ago. ;)

But when I buy a Macbook Pro 10,1, which I would like to do, I have to know what works and what not before buying :)

**Comment by gch1p** [comment]
> Is the sound on the MBA 5,2 headphone fixed/fixable?

Not sure right now, last time I rebased with master, the headphones bug was still there on MBA 5,2. But it might magically disappear if I rebase it with fresh master, as it happened with MBP 10,1. It was affected too, but then, I guess, some fixes to the azalia subsystem in coreboot (there were plenty) also fixed this bug. I can check it next week or so, if you'd like. 

> How about issues on the Macbook Pro 10,1? Are there any major?

Well, not all RAM variants are supported, you need to check that first. To support new ones, I need SPD dumps, which is easy.

I think thunderbolt controller will not work, but I don't have anything compatible to test it. And there's this [iMac port](https://review.coreboot.org/c/coreboot/+/38883) with comment:
> Cactus Ridge, the Thunderbolt controller. There isn't even a damn datasheet for this thing, so it's a literal black magic box. I mean, the package of this thing is black and square... Anyway, cursed sand.

If Angel says it doesn't work (under coreboot) on his iMac, I wouldn't expect it to work on MacBooks. And, as he said, there is no datasheet for the thing.

There was also a rare weird issue with fan spinning at full speed when it had no reason to, I believe it's something SMC related (one more proprietary Apple chip) and fixable, because 1) I wrote [this](https://github.com/gch1p/smctool) and 2) there is [this](https://github.com/acidanthera/VirtualSMC/tree/master/Docs) cool project with many SMC commands documented. (Well, I hope, if you can read fan speed with SMC, you can also set it, seems logical.)

> I will buy a MBA 4,2 soon, so I think that will work smoothly since the picky maintainers merged it a while ago. ;)

Ha-ha, it's not like that :D  I even forgot, it's funny you mentioned it. 

You see, when it was [merged](https://review.coreboot.org/c/coreboot/+/11840) in 2015, there was no code review _at all_. It was just autogenerated code (by `autoport` utility) and it got merged the next day. Nobody tested it, I'm 100% sure, because it didn't work, or at least the internal display couldn't work, because 4 years later I [discovered](https://review.coreboot.org/c/coreboot/+/32606) that it uses incorrect video port, lol. Then I [added](https://review.coreboot.org/c/coreboot/+/33103) support for [ACPI](https://review.coreboot.org/c/coreboot/+/33102) (lid open/closing, battery level reporting, AC charging reporting, all that laptop stuff), [CMOS](https://review.coreboot.org/c/coreboot/+/33045), [fixed](https://review.coreboot.org/c/coreboot/+/50272) its USB configuration and [made it clear](https://review.coreboot.org/c/coreboot/+/51531) that, actually, what was merged in 2015, supports only 1 of 6 possible RAM configurations and added ability to support more (if people will send SPD dumps to me). The last one still isn't merged, as I see, it just got +1 and nobody cares anymore.

Summing it up: in 2015, when there was zero code review, they merged broken, unusable code, without working display, without battery (ACPI) support, without correct USB ports config, and even that would boot only if you were lucky to have the laptop sold with one of six possible memory chips, and there wasn't even a note about it anywhere. But nobody knew that, or nobody cared, and despite it even got on [phoronix](https://www.phoronix.com/scan.php?page=news_item&px=Coreboot-SNB-MacBook-Air), totally broken mba4,2 code was lying in the repo there for 4 years. Now, in 2021, code review is so so frustrating that it's close to impossible to get merged a code that is tested, documented, where almost everything works, it's just total bureaucracy. Well, I agree that having code review is better then not having it, but this chicanery I can't stand anymore.

**Comment by cedricbambooza** [comment]
Oh G, that sounds terrible. 

Good to know that before buying a device :D 

So does the Thunderbolt work on any of the Macbooks? 

And if you could check on the MBA 5,2 headphones thing, when you find the time, I would really appreciate it. 

I also made a mistake. ^^ 
I thought the 10,1 was the 13" model, but since it's the 15" with the shitty descrete graphics, I'm not that interested in it. 
Even though iGPU works with your coreboot, it's simply the size I'm not that into then. 

I guess the MBA 5,2 or the MBP 8,1 will be the devices of my choice now. 
The status on those I can from your links, so that's pretty easy for me to decide what to choose, especially when the headphone thing on the MBA 5,2 might be fixed. 

What I didn't get clearly, is the MBA 4,2 working as of now? 
Or do I also have to pull one of your unmerged Requests. 

It's sad, that this project is that complicated and also, that they put broken code out in the master builds.
Somehow it's also a little funny, because it seems like it always happens that way in many places of Open Source Coding :D

**Comment by gch1p** [comment]
> So does the Thunderbolt work on any of the Macbooks?

I don't know. I think I can try to test it with a Thunderbolt-to-Ethernet adapter and see. But most probably not.

MBP 8,1 is, I think, the most bug-free and stable port, but if you will decide to buy it and flash externally, read this https://review.coreboot.org/c/coreboot/+/33151/16/Documentation/mainboard/apple/macbookpro8_1.md and see this photo https://review.coreboot.org/c/coreboot/+/33151/16/Documentation/mainboard/apple/mbp81_clip.jpg

There is a tiny resistor very close to the bios chip and it's easy to break with a clip, and I recommend to cut some plastic from a clip to be safe, as shown on the picture. Speaking from experience (I managed to break it once) :D

> What I didn't get clearly, is the MBA 4,2 working as of now?

I don't have 4,2 to test it, but some person said he failed to flash it internally, but external flashing worked and coreboot works too, linux boots. That's all I know. It also means that he was lucky to have that one of six possible memory chips that is supported.

About memory, in case you don't know: all these Air models, and Pro since the first retina model (which is 10,1), they have soldered RAM (well, this I'm sure you know). But, to make this soldered RAM work, BIOS has to know its parameters, think of it like "settings" for this RAM. BIOS stores all these "settings", checks via special GPIOs, which RAM is installed on the machine, and "loads" corresponding settings. According to schematics (hardware documentation), that is fortunately available for these MacBooks, they were sold with different RAM chips manufactured by different companies, such as Samsung, Hynix, etc. They all have different "settings". And, for example, when I added support for MBA 5,2, I could only add support for the memory that I have soldered in my specific laptop. To support all possible memory options, we need people who have macbooks with other memory chips to run a couple of programs under linux to get these "settings" (they're actually called SPD dumps).

As I remember, only 1 of 6 possible memory variants are supported for MBA 4,2 now, because we only have dump from one device.

**Comment by cedricbambooza** [comment]
Ok, but your patches seem to be in master. 
So I could try the coreboot on the MBA 4,2, since I'm lucky and it's 2021, not 2015 xD

I'll try to get one MBA 4,2 from ebay this month and then try the coreboot. 

If it works, I'm happy. 
If not, I will reflash stock bios and get you the SPD Dump plus the manufacturer of the Ram. 
In anyways it will be a 4G, I just can't tell what make of RAM I will get since ebay sellers are not that coorperative in that way, most of the time. 

`MBP 8,1 is, I think, the most bug-free and stable port, but if you will decide to buy it and flash externally, read this https://review.coreboot.org/c/coreboot/+/33151/16/Documentation/mainboard/apple/macbookpro8_1.md and see this photo https://review.coreboot.org/c/coreboot/+/33151/16/Documentation/mainboard/apple/mbp81_clip.jpg

There is a tiny resistor very close to the bios chip and it's easy to break with a clip, and I recommend to cut some plastic from a clip to be safe, as shown on the picture. Speaking from experience (I managed to break it once) :D` 

Thanks for all the hints :D

I definitly want to get a MBP 8,1 as well. 

For the moment it will only be the MBA 4,2, but the MBP 8,1 will follow. 

Ja the Thunderbolt Thing you don't have to try. 
I will test it on the go, since the hopes are not that high ;D

**Comment by cedricbambooza** [comment]
Got news on Thunderbolt 1 :) 

So it seems that the Kernel has driver support for Cactus Ridge, ABC Ridge  and XYZ Ridge ( made the last ones up ;D) 

Also by now for tunneling up to 6 devices. 

Question would be now, will Thunderbolt work through the Kernel, or can it be seen by it, when coreboot doesn't work with it? 

I'm not that deep into the Bios thing yet to know that.  

Maybe you have an answer to it. :)

**Comment by gch1p** [comment]
I'll check and let you know.

**Comment by cedricbambooza** [comment]
Thank you very much :)

**Comment by cedricbambooza** [comment]
So coreboot is not working, neither with SeaBios nor with Grub2. 
It spits a lot of Warnings during make process, all SeaBios or Grub2 related. 
When flashed, I get a black screen on the MBA4,2, but the fan is running normaly. 

I also can't try an older version, for some reason gcc 8.3 always fails to build.

**Comment by cedricbambooza** [comment]
Yes, that's SeaBios. 

I will try it with Grub once again, I did not mark the "Add Grub Config File to Rom" Option. :S 

But I have to create one first then, that's a bit of a downer.  

If it works or not, I will report.

**Comment by cedricbambooza** [comment]
The coreboot.cfg looks like this: 


[coreboot-cfg.txt](https://github.com/gch1p/mmga/files/7312885/coreboot-cfg.txt)

I did not run `grub-mkconfig` at any time of the process. 

I just build it from the folder and at the end of build, it said `boot/grub/grub.cfg = coreboot.cfg`, so I figured that I don't have to do something else. 

During build of coreboot, it also didn't complain about the .cfg file.
